baseline-network-linux-hld.md

```diff
@@ -0,0 +1,250 @@
+building and development on-device. Each OS specific build system should also be
+able to consume and utilise the shared platform support sources from this
+independently managed repository.
+
+Any packaging decisions like if or and how to split up individual platform
+support should be up to the OS distribution packaging it.
+
+This allows non-Yocto based Operation Systems to make use of the platform
+support.
+
+### Base platform support Yocto layer (BSP layer)
+
+The base platform support Yocto layer will provide the appropriate recipes and
+definitions for running on supported platforms. It will contain recipes for the
+kernel, the platform support, and appropriate machine definitions to target
+platforms based on their architecture.
+
+### Yocto based example NOS (Distribution layer)
+
+To allow quick testing and easy extension for other users, an example
+implementation and support layer for Yocto will be provided. It will contain all
+basic building blocks for creating an installable image, i.e. a distribution
+definition, as well as appropriate recipes for creating ONIE installable images.
+
+## Functional Description
+
+### Base Open Networking Hardware Support
+
+The base networking hardware support combines hardware support from multiple
+sources, and therefore will need to support multiple kinds of APIs.
+
+Possible APIs are:
+
+- SONiC PDDF, due to the large amount of available platforms
+  https://github.com/sonic-net/SONiC/blob/master/doc/platform/brcm_pdk_pddf.md
+- S3IP as described in https://github.com/sonic-net/SONiC/pull/1068
+- ONLP, as used by ONL and DentOS
+
+The base networking hardware support is organized in multiple repositories:
+
+- kernel: Linux fork with required base changes (backports of important
+  submissions, etc)
+- hardware-platform-support: drivers, libraries and userspace code for
+  initializing hardware platforms (do we need more?)
+- TBD: organization of platform support
+- TBD: platform initialization
+
+For the first iteration, the platform support code can be left as is
+in their current source repositories.
+
+Future iterations should focus on improving the platform drivers and support
+code to follow modern code style and compile with as few warnings as possible,
+ideally none.
+
+New platform support should follow modern coding styles from the beginning (e.g.
+kernel coding style for modules).
+
+### Kernel recipe and configuration handling
+
+The kernel configuration will be done via
+https://docs.yoctoproject.org/kernel-dev/advanced.html to allow large
+flexibility in how other projects using the layer can customize the kernel.
+
+TBD: should we have patches or branches with backport commits? Maybe both?
+
+- patches make required changes due to upstream changes more obvious (provide a
+  “conflict diff”) and easier to review
+- branches are closer to the upstream approach, but may be harder to track
+  changes needed due to upstream changes
```

Comment:
I'd suggest maintaining separate feature branches and rebase them on each new upstream kernel change. This way, each commit in a topic branch can be submitted upstream as soon as possible and the remaining downstream changes are obvious.

A rebase can be reviewed easily by using `git range-diff`.

Reply:
My reviewability concern was about reviewing pull requests or submitted patches - you can't submit a rebase for being reviewed by others.

So basically AFAICT if you want the rebase to a newer kernel reviewable, you need to create a pull request for the changes (or submit a patchset), then compare the diff with the diff of the old branch to ensure no important changes were dropped, or wrong rebase decisions were taken in merge conflicts.

A patch updating patches makes this more obvious to me, but then again I have an OpenWrt background, so I might be a bit biased there ;-)

Reply:
> My reviewability concern was about reviewing pull requests or submitted patches - you can't submit a rebase for being reviewed by others.
> 
> So AFAICT if you want the rebase to a newer kernel reviewable, you need to create a pull request for the changes (or submit a patchset), then compare the diff with the diff of the old branch to ensure no important changes were dropped, or wrong rebase decisions were taken in merge conflicts.

In my experience, GitHub reviews work well for incremental changes being merged into the main branch (if the PR is targeted at the upstream repo). But for kernel patches, we have a different situation: While they are used in a downstream kernel (as in this case), they need to be rebased again and again.

When merging in new upstream releases (as in https://github.com/dentproject/linux/pull/4) one after the other, cherry-picking any commit for upstreaming becomes harder and harder because the necessary changes are hidden in the merge commits.

So the only sustainable way is to keep rebasing the set of out-of-tree patches for each upstream release, thereby updating each patch to be compatible again. Ideally as topic branches, to avoid introducing unnecessary dependencies between patches (which would make upstreaming harder).

But if you keep these only as a set of patches in git, you don't have the necessary context during PR review when looking at diffs of patches on that repository. Renames or removals of patches easily create so much noise that the real differences are hidden.

At my day-job, we use [umpf](https://github.com/pengutronix/umpf) to build a linear series (in git and/or as patches) from the topic branches. When updating the combined patch series in a BSP layer, we then add the `range-diff` as a comment to the PR, as that's much easier to review. The individual topic branch rebases can be reviewed with `range-diffs` outside of PRs if needed. There's a [blog post](https://pengutronix.de/en/blog/2023-08-29.html) with more background. I can't guarantee the tool would be a good fit here, but I think one should take a look. ;)

> A patch updating patches makes this more obvious to me, but then again I have an OpenWrt background, so I might be a bit biased there ;-)

:)

Reply:
> > My reviewability concern was about reviewing pull requests or submitted patches - you can't submit a rebase for being reviewed by others.
> > So AFAICT if you want the rebase to a newer kernel reviewable, you need to create a pull request for the changes (or submit a patchset), then compare the diff with the diff of the old branch to ensure no important changes were dropped, or wrong rebase decisions were taken in merge conflicts.
> 
> In my experience, GitHub reviews work well for incremental changes being merged into the main branch (if the PR is targeted at the upstream repo). But for kernel patches, we have a different situation: While they are used in a downstream kernel (as in this case), they need to be rebased again and again.
> 
> When merging in new upstream releases (as in [dentproject/linux#4](https://github.com/dentproject/linux/pull/4)) one after the other, cherry-picking any commit for upstreaming becomes harder and harder because the necessary changes are hidden in the merge commits.
> 
> So the only sustainable way is to keep rebasing the set of out-of-tree patches for each upstream release, thereby updating each patch to be compatible again. Ideally as topic branches, to avoid introducing unnecessary dependencies between patches (which would make upstreaming harder).
> 
> But if you keep these only as a set of patches in git, you don't have the necessary context during PR review when looking at diffs of patches on that repository. Renames or removals of patches easily create so much noise that the real differences are hidden.

> At my day-job, we use [umpf](https://github.com/pengutronix/umpf) to build a linear series (in git and/or as patches) from the topic branches. When updating the combined patch series in a BSP layer, we then add the `range-diff` as a comment to the PR, as that's much easier to review. The individual topic branch rebases can be reviewed with `range-diffs` outside of PRs if needed. There's a [blog post](https://pengutronix.de/en/blog/2023-08-29.html) with more background. I can't guarantee the tool would be a good fit here, but I think one should take a look. ;)

Interesting tool, I will need to take a closer look at it.

I guess my concern with rebasing the out-of-tree changes into a new branch (and then updating the recipe to build from the new branch/commit) is that this doubles the work needed for updating the kernel - once for the new branch, then for updating the recipe.

This also has the side effect that all pending PRs for the kernel will now have to be retargeted to the new branch, which can be painful, depending on how good you are with git.

But hopefully we won't have too many local changes, and push hard for upstream first, even for backporting fixes to stable versions (kernel maintainers are also only human, and easily overlook fixes worth backporting).